### PR TITLE
Lightweight application offers.

### PR DESCRIPTION
**Pull request by alesstimec**
## Description

    Removes data from ApplicationOffer type that JIMM does not need to store. Instead the source of
    truth for application offers becomes the offering juju controller. Controllers are now queries when
    we need details on application offers.
    
    It also allowed me to remove the update method for application offers and simplify the
    AllWatcher so that we don't have to process deltas for offers.
    
    A drive-by fix: we no longer require the user to have add-model permission to add a k8s cloud.

## Engineering checklist
*Check only items that apply*

- [ ] Documentation updated
- [ ] Covered by unit tests
- [ ] Covered by integration tests

### CODE REVIEW

**Comment by kian99** [comment]
I think we have a solution #1441, let's wait for that to land before landing this please.

File: internal/jimm/applicationoffer.go
Comment (kian99) [inline internal/jimm/applicationoffer.go]: I like this change of just querying the Juju controllers but we have some challenges with this approach:
1. It will query all offers on all controllers each time. That could be a large number of offers.
2. It doesn't take into account the fact that model owners have changed when a model was imported into JIMM.
3. It doesn't take into account the difference between `juju find-offers` and `juju offers`. The latter shows you the offers available in a model. The former shows you offers you have access to from all models so that you can use them in your current model.

I think we can solve this pretty easily with some code that already existed. List-offers should just list the offers in the specified model as was already done, no problems there. Find-offers can loop through all controllers/offers but requires the model owner to be treated carefully. If the model owner/name are specified we can just use the approach in list-offers.

Alternatively, we can consider removing the transfer of ownership concept when importing models into JIMM. That would mean two models with the same name/owner on different controllers can't be imported into JIMM but that could possibly be an okay tradeoff.

Comment (alesstimec) [reply internal/jimm/applicationoffer.go]: @kian99 updated as you suggested:
- ListOffers now works as before
- FindApplicationOffers will query all controllers in parallel - not the most efficient, but better than the serial approach
- still does not handle the case of imported models

File: internal/db/applicationoffer.go
Comment (kian99) [inline internal/db/applicationoffer.go:109]: How does this interact with the preload below? Or rather do we need the preload?

File: internal/jimm/applicationoffer.go
Comment (kian99) [inline internal/jimm/applicationoffer.go]: What do you think about using the `enrichOfferDetails` function here and elsewhere where we do this?

**Review by kian99 — COMMENTED** [review]
Looks good, just 2 comments so far. Also can we please update the PR description

File: internal/db/applicationoffer.go
Comment (ale8k) [inline internal/db/applicationoffer.go:90]: Nice

File: internal/db/applicationoffer.go
Comment (ale8k) [reply internal/db/applicationoffer.go:109]: Yeah I'm confused here too, will we need to access the model within the offers on this db call?

File: internal/dbmodel/sql/postgres/1_14.sql
Comment (ale8k) [inline internal/dbmodel/sql/postgres/1_14.sql:2]: Ooft... This definitely felt good

File: internal/jimm/watcher.go
Comment (ale8k) [inline internal/jimm/watcher.go:496]: Default and do what?

File: internal/jujuapi/applicationoffers_test.go
Comment (ale8k) [inline internal/jujuapi/applicationoffers_test.go]: I thought we were not going to in line func calls i.e., from our discussion on c.Assert(mythingreturnserr, gc.IsNil)? This seems like the same thing?

File: internal/jujuapi/applicationoffers_test.go
Comment (ale8k) [inline internal/jujuapi/applicationoffers_test.go]: Learning question (maybe I've forgotten actually): What is this line? "NoRelation"?

**Review by ale8k — APPROVED** [review]
Couple comments, I'm approving but ping me when responded.

File: internal/jimm/watcher.go
Comment (alesstimec) [reply internal/jimm/watcher.go:496]: nothing.. ignore deltas we don't care about

File: internal/jujuapi/applicationoffers_test.go
Comment (alesstimec) [reply internal/jujuapi/applicationoffers_test.go]: asserting user does not have access to the application offer.

File: internal/db/applicationoffer.go
Comment (alesstimec) [reply internal/db/applicationoffer.go:109]: a bit of shameless cargo-culting from cloud defaults. Joins and preloads do not seem to interfere with eachother.

File: internal/jimm/applicationoffer.go
Comment (alesstimec) [reply internal/jimm/applicationoffer.go]: done.

Comment (alesstimec) [reply internal/jimm/applicationoffer.go]: actually.. done with a little tweak

File: internal/jujuapi/applicationoffers_test.go
Comment (alesstimec) [reply internal/jujuapi/applicationoffers_test.go]: fixed

File: internal/db/applicationoffer.go
Comment (ale8k) [reply internal/db/applicationoffer.go:109]: Yeah it'll filter before actually loading I imagine

File: internal/jimm/applicationoffer.go
Comment (kian99) [inline internal/jimm/applicationoffer.go]: nit
```suggestion
// where JIMM is authoritative. It returns a noApplicationOfferAccessError if the user
```

File: internal/jimm/applicationoffer.go
Comment (kian99) [inline internal/jimm/applicationoffer.go]: I wonder if we shouldn't just query all controllers after reading this.
It's either:
1. Query OpenFGA for all offers a user can access.
2. Query DB for each offer to get the controller of the offer.
3. Query the controllers with the user provided filters.
4. Check the user has access to the returned offers.

Versus

1. Query all controllers with the user provided filters.
2. Check the user has access to the returned offers.

File: internal/jimm/cloud.go
Comment (kian99) [inline internal/jimm/cloud.go:227]: Should've been a separate PR.

File: internal/jimm/applicationoffer.go
Comment (kian99) [reply internal/jimm/applicationoffer.go]: It's worth it to keep `ListApplicationOffers` as is since that is just querying a single model. But I would simplify `FindApplicationOffers` to just query across all the controllers with the provided filters.

File: internal/jimm/cloud.go
Comment (alesstimec) [reply internal/jimm/cloud.go:227]: yes, it should, but i (we) keep forgetting about it :)

File: internal/jimm/applicationoffer.go
Comment (alesstimec) [reply internal/jimm/applicationoffer.go]: done.

File: internal/jimm/applicationoffer.go
Comment (kian99) [inline internal/jimm/applicationoffer.go]: Minor optimisation for a follow up, but instead of passing in the controller ID you could pass in the controller dbmodel since we fetch it twice.

File: internal/jimm/applicationoffer.go
Comment (kian99) [inline internal/jimm/applicationoffer.go]: You don't need to do this since Go 1.23

File: internal/jimm/applicationoffer.go
Comment (kian99) [inline internal/jimm/applicationoffer.go]: The `dbOffer` parameter is no longer needed in this function.

**Review by kian99 — APPROVED** [review]
lgtm with some comments, happy for them to get done in a follow up.

**Pull request merged**
